rfcs/0024-python-scope-package-set.md

```diff
@@ -0,0 +1,125 @@
+Python packages that are used in a development environment. This RFC formalizes
+and thereby clarifies this distinction and defines as well the scope of the
+*package set*, which is a curated set of compatible libraries that can be used
+in a development environment.
+
+# Motivation
+[motivation]: #motivation
+
+The motivation of this RFC is to formalize and thereby clarify the current way
+of working regarding the package set scope. The reason to have this formalized
+is to have a clear reference and guideline when making the daily decisions about
+what to include or exclude from the package set.
+
+# Detailed design
+[design]: #detailed-design
+
+## Introduction
+
+The Nix Packages collection (Nixpkgs) consists of a large amount of expressions
+of which a significant portion corresponds to package sets. Major package sets
+at the time of writing exist e.g. for the Haskell, Perl and Python languages.
+The scope and way working with these package sets depends on multiple factors,
+such as the tools provided upstream, the demand from the community and the
+maintainers of the package sets and the individual expressions. Another
+important factor is whether *upstream* provides a curated set of packages that
+are known to be compatible. An example of a curated set or snapshot is Stackage
+for Haskell. If such set exists and serves the needs from the community, then
+the scope may be limited to providing that set.
+
+## Definitions
+
+The following definitions are used in the scope of this RFC:
+- application: a package providing programs that happen to be written in Python
+- library: a package providing modules and programs that are used in a Python environment for development purposes or as dependency for applications
+- package set: a curated set of compatible libraries that can be used in a development environment and as dependencies for applications
+- maintainer: a maintainer of an expression as typically noted in the `meta.maintainers` field
+- package set maintainer: a maintainer of the Python package set as noted in the `CODEOWNERS` file
+
+Examples on the usage of application and library:
+- `scipy` is a collection of routines for scientific computing. It is used for development purposes or as a library and is therefore considered a *library*
+- `calibre` is a program for managing e-books that happens to be written in Python. The package is therefore considered an *application*
+- `pytest` is a test runner. While a program, it can only be used in conjunction with the Python environment where it needs to import the modules that are to be tested. It is therefore considered a *library*
+
+## Scope of the package set
+
+The purpose of the package set is to provide a set of compatible libraries that
+can be used for development purposes and as dependencies by Python applications.
+Applications are not considered part of the package set and their expressions
+shall thus also be placed elsewhere.
```

Comment:
Yep, fixed in 473c470. That will require a new function `toPythonApplication` which basically does the opposite of the current `toPythonModule`.

Reply:
Are there any real pure-Python applications that _cannot_ be imported as a library?

Reply:
If they don't put their modules in `site-packages`, then it gets hard to import them. So, if they don't use setuptools/distutils/flit, that typically is the case.

However, I don't think it is important whether it is technically possible, but whether it is meant to. I would argue that upstream application developers typically use `setuptools` to make it easier to distribute and that it is therefore just an implementation detail.

Reply:
I agree with your examples where there is a technical distinction. But what is gained by relying on intent when there is no technical hard distinction? If it is built just like a Python library, and it can be used as a Python library, and it can be generated by `pypi2nix` and copied into the package set with a minimal cleanup — shouldn't it be duck-typed to a «Python package» and then maybe also aliased to the top-level namespace?

Reply:
> But what is gained by relying on intent when there is no technical hard distinction?

No practical gain. But as the language is an implementation detail it could be changed, whereas its purpose and thus its location in the tree would remain the same. Although, as you write, a pragmatic solution could be:

> and then maybe also aliased to the top-level namespace?

In practice, though, there is actually another reason to have them separate: the additional constraints (pinning) typically imposed by applications (https://github.com/NixOS/nixpkgs/issues/34775).

Reply:
In my experience, libraries can also have interesting version constraints.

I would call a complete rewrite, same or different language, a new package by the same team. Also, of packages I use, Mozilla Marionette (which is a Python package `pythonPackages.marionette_driver` with no executables — so a library) still supports only Python2 and I wouldn't be surprised if it gets rewritten in Rust instead of updating to Python3.